File: go/todo_test.go
package spec

import (
	"encoding/json"
	"testing"
)

func Test_TodoPart_construct_and_decode(t *testing.T) {
	part := NewTodoPart(TodoPart{
		ID:    "todo_main",
		Title: "Release",
		Items: []TodoItem{
			{ID: "t1", Content: "Port codec", Status: TodoCompleted, ActiveForm: "Porting codec"},
			{ID: "t2", Content: "Wire sahara", Status: TodoInProgress},
		},
	})
	if part.Type() != PartTodo {
		t.Fatalf("type = %q", part.Type())
	}
	body, ok := part.AsTodo()
	if !ok {
		t.Fatalf("AsTodo failed")
	}
	if body.ID != "todo_main" || len(body.Items) != 2 {
		t.Fatalf("body = %#v", body)
	}
	if body.Items[0].Status != TodoCompleted || body.Items[1].Status != TodoInProgress {
		t.Fatalf("statuses = %v / %v", body.Items[0].Status, body.Items[1].Status)
	}
	if body.Items[0].ActiveForm != "Porting codec" {
		t.Fatalf("active_form lost: %#v", body.Items[0])
	}
}

func Test_TodoPart_survives_memorylayer_codec(t *testing.T) {
	part := NewTodoPart(TodoPart{ID: "todo_main", Items: []TodoItem{
		{ID: "t1", Content: "A", Status: TodoInProgress},
	}})
	msg := ChatMessage{SchemaVersion: "1.0", ID: "m1", Role: RoleAssistant, Addr: MessageAddress{ThreadID: "thr"}, Content: []ContentPart{part}}

	payload, err := ToMemoryLayerPayload(msg)
	if err != nil {
		t.Fatalf("to payload: %v", err)
	}
	raw, _ := json.Marshal(map[string]any{
		"id": "ml", "thread_id": "thr", "role": payload["role"],
		"content": payload["content"], "metadata": payload["metadata"],
	})
	var generic map[string]any
	if err := json.Unmarshal(raw, &generic); err != nil {
		t.Fatalf("unmarshal record: %v", err)
	}
	back, err := FromMemoryLayerMessage(generic)
	if err != nil {
		t.Fatalf("from message: %v", err)
	}
	if len(back.Content) != 1 || back.Content[0].Type() != PartTodo {
		t.Fatalf("codec lost todo part: %#v", back.Content)
	}
	rt, ok := back.Content[0].AsTodo()
	if !ok || len(rt.Items) != 1 || rt.Items[0].Status != TodoInProgress {
		t.Fatalf("todo items lost through codec: %#v", rt)
	}
}

File: typescript/test/todo.test.ts
/**
 * Tests for the todo content part — shared task tracking.
 *
 * Mirrors go/todo_test.go and python/tests/test_todo.py, and exercises the live
 * update path (part_appended → part_updated) through the reducer.
 */
import {describe, expect, it} from 'vitest';
import {
    fromMemoryLayerMessage,
    reduceEvents,
    toMemoryLayerPayload,
    type ChatMessage,
    type MessageStartedEvent,
    type PartAppendedEvent,
    type PartUpdatedEvent,
    type TodoItem,
    type TodoPart,
} from '../src/index';

function todoPart(items: TodoItem[]): TodoPart {
    return {type: 'todo', id: 'todo_main', items};
}

describe('todo part', () => {
    it('streams live: part_appended then part_updated replaces items via shallow merge', () => {
        const msg: ChatMessage = {
            schema_version: '1.0', id: 'm1', role: 'assistant',
            addr: {thread_id: 'thr'}, meta: {}, content: [],
        };
        const state = reduceEvents([
            {event: 'message_started', message: msg} as MessageStartedEvent,
            {
                event: 'part_appended', message_id: 'm1', index: 0,
                part: todoPart([{id: 't1', content: 'A', status: 'pending'}]),
            } as PartAppendedEvent,
            {
                event: 'part_updated', message_id: 'm1', index: 0,
                patch: {items: [
                    {id: 't1', content: 'A', status: 'completed'},
                    {id: 't2', content: 'B', status: 'in_progress'},
                ]},
            } as PartUpdatedEvent,
        ]);
        const part = state['m1']!.content[0] as TodoPart;
        expect(part.type).toBe('todo');
        expect(part.items).toHaveLength(2);
        expect(part.items[0]!.status).toBe('completed');
        expect(part.items[1]!.status).toBe('in_progress');
    });

    it('round-trips through the MemoryLayer codec', () => {
        const msg: ChatMessage = {
            schema_version: '1.0', id: 'm1', role: 'assistant',
            addr: {thread_id: 'thr'}, meta: {},
            content: [todoPart([{id: 't1', content: 'A', status: 'in_progress', active_form: 'Doing A'}])],
        };
        const payload = toMemoryLayerPayload(msg);
        const rec = JSON.parse(JSON.stringify({
            id: 'ml', thread_id: 'thr', role: payload.role,
            content: payload.content, metadata: payload.metadata,
        }));
        const back = fromMemoryLayerMessage(rec);
        const part = back.content[0] as TodoPart;
        expect(part.type).toBe('todo');
        expect(part.items[0]!.status).toBe('in_progress');
        expect(part.items[0]!.active_form).toBe('Doing A');
    });
});
